File: core/java/src/main/java/hara/lang/data/MapEntry.java
package hara.lang.data;

import hara.lang.base.Ex;
import hara.lang.base.Eq;
import hara.lang.base.G;
import hara.lang.base.Iter;
import hara.lang.data.types.ObjPersistent;
import hara.lang.protocol.Constant;
import hara.lang.protocol.ICount;
import hara.lang.protocol.IEquality;
import hara.lang.protocol.IMetadata;
import hara.lang.protocol.INth;
import hara.lang.protocol.IPair;
import java.util.Iterator;

/** The immutable, pair-only representation of one map entry. */
public final class MapEntry<K, V> extends ObjPersistent
    implements IPair<K, V>, ICount, INth<Object>, IEquality, Iterable<Object> {
  private final K key;
  private final V value;

  public MapEntry(IMetadata meta, K key, V value) {
    super(meta);
    this.key = key;
    this.value = value;
  }

  @Override
  public K getKey() {
    return key;
  }

  @Override
  public V getValue() {
    return value;
  }

  @Override
  public long count() {
    return 2;
  }

  @Override
  public Object nth(long index) {
    if (index == 0) return key;
    if (index == 1) return value;
    throw new Ex.NoSuchElement();
  }

  @Override
  public Iterator<Object> iterator() {
    return Iter.objects(key, value);
  }

  @Override
  public MapEntry<K, V> withMeta(IMetadata meta) {
    return _meta == meta ? this : new MapEntry<>(meta, key, value);
  }

  @Override
  public Constant.ObjType getObjType() {
    return Constant.ObjType.MAP_ENTRY;
  }

  /** Map entries retain the ordered two-value hash used by sequential tuples. */
  @Override
  public String hashSeed() {
    return "::SEQUENTIAL";
  }

  @Override
  public long hashCalc(Constant.HashType type) {
    return Iter.reduce(
        iterator(),
        Long.valueOf(hashSeed().hashCode()),
        (acc, item) -> (acc * 31) + G.hashFn(type).apply(item));
  }

  @Override
  public boolean equality(Object other) {
    return other instanceof MapEntry<?, ?> entry
        && Eq.eq(key, entry.getKey())
        && Eq.eq(value, entry.getValue());
  }

  @Override
  public boolean equals(Object other) {
    return equality(other);
  }

  @Override
  public int hashCode() {
    return Long.hashCode(hashCalc(Constant.HashType.RAPID));
  }

  @Override
  public String display() {
    return "[" + G.display(key) + " " + G.display(value) + "]";
  }
}

File: core/rust/src/lang/data/map_entry.rs
use crate::core::Value;
use crate::lang::hash::JavaHash;
use crate::lang::protocol::{
    HashType, ICount, IDisplay, IEquality, IHash, IMetadata, INth, IObjType, IPair, ObjType,
};
use std::rc::Rc;

/// The immutable runtime representation of a map entry.
///
/// Compact tuples deliberately remain a separate internal representation for
/// small vectors. A `MapEntry` is the only value that implements `IPair`.
#[derive(Debug, Clone)]
pub struct MapEntry {
    key: Value,
    value: Value,
    metadata: Option<Rc<crate::lang::data::Metadata>>,
}

impl MapEntry {
    pub fn new(key: Value, value: Value) -> Self {
        Self {
            key,
            value,
            metadata: None,
        }
    }

    pub fn key(&self) -> &Value {
        &self.key
    }

    pub fn value(&self) -> &Value {
        &self.value
    }

    pub fn nth(&self, index: usize) -> Option<&Value> {
        match index {
            0 => Some(&self.key),
            1 => Some(&self.value),
            _ => None,
        }
    }

    pub fn iter(&self) -> std::array::IntoIter<&Value, 2> {
        [&self.key, &self.value].into_iter()
    }
}

impl PartialEq for MapEntry {
    fn eq(&self, other: &Self) -> bool {
        self.key == other.key && self.value == other.value
    }
}

impl Eq for MapEntry {}

impl ICount for MapEntry {
    fn count(&self) -> usize {
        2
    }
}

impl INth<Value> for MapEntry {
    fn nth(&self, index: usize) -> Option<&Value> {
        self.nth(index)
    }
}

impl IPair<Value, Value> for MapEntry {
    fn key(&self) -> &Value {
        self.key()
    }

    fn value(&self) -> &Value {
        self.value()
    }
}

impl IMetadata for MapEntry {
    type Metadata = Rc<crate::lang::data::Metadata>;

    fn meta(&self) -> Option<&Self::Metadata> {
        self.metadata.as_ref()
    }

    fn with_meta(&self, metadata: Option<Self::Metadata>) -> Self {
        Self {
            key: self.key.clone(),
            value: self.value.clone(),
            metadata,
        }
    }
}

impl IEquality for MapEntry {
    fn equality(&self, other: &Self) -> bool {
        self == other
    }
}

impl IDisplay for MapEntry {
    fn display(&self) -> String {
        format!("[{} {}]", self.key.display(), self.value.display())
    }
}

impl IHash for MapEntry {
    fn hash_calc(&self, hash_type: HashType) -> u64 {
        crate::lang::hash::compose_ordered(
            "SEQUENTIAL",
            self.iter().map(|value| value.java_hash(hash_type)),
        ) as u64
    }
}

impl IObjType for MapEntry {
    fn obj_type(&self) -> ObjType {
        ObjType::MapEntry
    }

    fn hash_seed(&self) -> String {
        "::SEQUENTIAL".into()
    }
}
